File: sth/java/sort/qs.java
class Solution {
    public void sortColors(int[] nums) {
        int n = nums.length - 1;
        qs(nums, 0, n);   
    }
    
    private void qs(int[] nums, int p, int r){
        if (p < r){
            int q = patition(nums, p, r);
            qs(nums, p, q-1);
            qs(nums, q+1, r);
        }
    }
    
    private int patition(int[] nums, int p, int r){
        System.out.print(r);
        int x = nums[r];
        int i = p-1;
        for (int j=p; j<r; j++){
            if (nums[j] < x){
                i = i + 1;
                swap(nums, i, j);
            }    
        }
        swap(nums, i+1, r);
        return i+1;
    }
    
    private void swap(int[] nums, int a, int b){
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
}

File: sth/go/sort/qs.go
func sortColors(nums []int)  {
    n := len(nums) -1 
    qs(nums, 0, n)
    
}

func qs(nums []int, p, r int) {
    if p < r {
        q := partition(nums, p, r)
        qs(nums, p, q-1)
        qs(nums, q+1, r)
    }
}

func partition(nums []int, p, r int) int {
    x := nums[r]
    i := p-1
    
    for j:=p; j<r; j++ {
        if nums[j] < x {
            i++
            nums[i], nums[j] = nums[j], nums[i]
        }
    }
    nums[i+1], nums[r] = nums[r], nums[i+1]
    return i+1
    
}
